cranelift-object/src/backend.rs

```diff
@@ -0,0 +1,561 @@
+            symbol.scope = scope;
+            symbol.weak = weak;
+        } else {
+            let symbol_id = self.object.add_symbol(Symbol {
+                name: name.as_bytes().to_vec(),
+                value: 0,
+                size: 0,
+                kind: SymbolKind::Text,
+                scope,
+                weak,
+                section: None,
+            });
+            self.functions[id] = Some(symbol_id);
+        }
+    }
+
+    fn declare_data(
+        &mut self,
+        id: DataId,
+        name: &str,
+        linkage: Linkage,
+        _writable: bool,
+        _align: Option<u8>,
+    ) {
+        let scope = match linkage {
+            Linkage::Import => SymbolScope::Unknown,
+            Linkage::Local => SymbolScope::Compilation,
+            Linkage::Export | Linkage::Preemptible => SymbolScope::Dynamic,
+        };
+        // TODO: this matches rustc_codegen_cranelift, but may be wrong.
+        let weak = linkage == Linkage::Preemptible;
+
+        if let Some(data) = self.data_objects[id] {
+            let symbol = self.object.symbol_mut(data);
+            symbol.scope = scope;
+            symbol.weak = weak;
+        } else {
+            let symbol_id = self.object.add_symbol(Symbol {
+                name: name.as_bytes().to_vec(),
+                value: 0,
+                size: 0,
+                kind: SymbolKind::Data,
+                scope,
+                weak,
+                section: None,
+            });
+            self.data_objects[id] = Some(symbol_id);
+        }
+    }
+
+    fn define_function(
+        &mut self,
+        func_id: FuncId,
+        _name: &str,
+        ctx: &cranelift_codegen::Context,
+        _namespace: &ModuleNamespace<Self>,
+        code_size: u32,
+    ) -> ModuleResult<ObjectCompiledFunction> {
+        let mut code: Vec<u8> = vec![0; code_size as usize];
+        let mut reloc_sink = ObjectRelocSink::default();
+        let mut trap_sink = ObjectTrapSink::default();
+        let mut stackmap_sink = NullStackmapSink {};
+
+        if let ObjectTrapCollection::Enabled = self.collect_traps {
+            unsafe {
+                ctx.emit_to_memory(
+                    &*self.isa,
+                    code.as_mut_ptr(),
+                    &mut reloc_sink,
+                    &mut trap_sink,
+                    &mut stackmap_sink,
+                )
+            };
+        } else {
+            let mut trap_sink = NullTrapSink {};
+            unsafe {
+                ctx.emit_to_memory(
+                    &*self.isa,
+                    code.as_mut_ptr(),
+                    &mut reloc_sink,
+                    &mut trap_sink,
+                    &mut stackmap_sink,
+                )
+            };
+        }
+
+        let symbol = self.functions[func_id].unwrap();
+        let section = self.object.section_id(StandardSection::Text);
+        let offset = self.object.add_symbol_data(symbol, section, &code, 1);
```

Comment:
Faerie uses an alignment of 0x10.

Reply:
Right, but it's a magic number with no documentation as to why it was selected. I prefer to stick to the minimum without good reason. My opinion is that this should be something the caller is required to select, or at least it should be an architecture dependent value (see #758). Same for the data alignment.

Reply:
`cranelift-object` doesn't provide any way to select the code alignment.

Reply:
I think I could change `cranelift-module` to allow that.

Reply:
I added an `ObjectBuilder` setting since you'll probably want the same alignment for all functions.